util/src/main/java/org/datacommons/util/ReconClient.java

```diff
@@ -22,30 +27,99 @@ public class ReconClient {
 
+  private static final String RESOLVE_ENTITIES_API_URL =
+      "https://api.datacommons.org/v1/recon/entity/resolve";
+
+  static final String NUM_API_CALLS_COUNTER = "ReconClient_NumApiCalls";
+
+  private static final int DEFAULT_CHUNK_SIZE = 500;
+
+  private final int chunkSize;
+
   private final HttpClient httpClient;
 
-  public ReconClient(HttpClient httpClient) {
+  private final LogWrapper logWrapper;
+
+  public ReconClient(HttpClient httpClient, LogWrapper logWrapper) {
+    this(httpClient, logWrapper, DEFAULT_CHUNK_SIZE);
+  }
+
+  public ReconClient(HttpClient httpClient, LogWrapper logWrapper, int chunkSize) {
     this.httpClient = httpClient;
+    this.logWrapper = logWrapper;
+    this.chunkSize = chunkSize;
   }
 
-  public ResolveCoordinatesResponse resolveCoordinates(ResolveCoordinatesRequest request)
-      throws IOException, InterruptedException {
-    return callApi(
-        RESOLVE_COORDINATES_API_URL, request, ResolveCoordinatesResponse.getDefaultInstance());
+  public CompletableFuture<ResolveCoordinatesResponse> resolveCoordinates(
+      ResolveCoordinatesRequest request) {
+    ResolveCoordinatesResponse defaultResponse = ResolveCoordinatesResponse.getDefaultInstance();
+    if (request.getCoordinatesCount() < 1) {
+      return CompletableFuture.completedFuture(defaultResponse);
+    }
+
+    return toFutureOfList(
+            partition(request.getCoordinatesList(), chunkSize).stream()
+                .map(
+                    chunk ->
+                        request.toBuilder().clearCoordinates().addAllCoordinates(chunk).build())
+                .map(
+                    chunkedRequest ->
+                        callApi(RESOLVE_COORDINATES_API_URL, chunkedRequest, defaultResponse))
+                .collect(toList()))
+        .thenApply(
+            chunkedResponses ->
+                ResolveCoordinatesResponse.newBuilder()
+                    .addAllPlaceCoordinates(
+                        chunkedResponses.stream()
+                            .flatMap(
+                                chunkedResponse ->
+                                    chunkedResponse.getPlaceCoordinatesList().stream())
+                            .collect(toList()))
+                    .build());
+  }
+
+  public CompletableFuture<ResolveEntitiesResponse> resolveEntities(
+      ResolveEntitiesRequest request) {
+    ResolveEntitiesResponse defaultResponse = ResolveEntitiesResponse.getDefaultInstance();
+    if (request.getEntitiesCount() < 1) {
+      return CompletableFuture.completedFuture(defaultResponse);
+    }
+
+    return toFutureOfList(
+            partition(request.getEntitiesList(), chunkSize).stream()
+                .map(chunk -> request.toBuilder().clearEntities().addAllEntities(chunk).build())
+                .map(
+                    chunkedRequest ->
+                        callApi(RESOLVE_ENTITIES_API_URL, chunkedRequest, defaultResponse))
+                .collect(toList()))
```

Comment:
I'm a bit unsure about the logic here.  Partly my ignorance in having never dealt with futures, and partly because its unclear to me whether doing doing this sort of async thing is warranted?   Worth also noting that these functions will be called concurrently, and might need to reason about that too.

So generally I wonder why not keep the logic addition separate from optimizations that change the threading/synchrony?  Get the basic lat/lng thing working e2e, measure perf etc, then do the refactors / optimizations.

Reply:
It's not about optimization but simplicity. What makes concurrent programming complicated is not parallelism but mutation. Mutation requires management - mutexes, locks, mutual exclusion. Here we are dealing with protos which are immutable and hence it is inherently concurrent. It just turns out that this simplicity also enables parallelism.

Anyway, I'm fine going with whichever approach you feel is the right thing to do. Let's discuss in office and go with that.